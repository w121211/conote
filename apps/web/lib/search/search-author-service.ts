import Fuse from 'fuse.js'
import { Author } from '@prisma/client'
import { AuthorModel } from '../models/author-model'

export class SearchAuthorServiceClass {
  private authors: Author[] | null = null // TODO: store in redis instead
  private fuse: Fuse<Author> | null = null

  add(author: Author): Fuse<Author> {
    if (this.fuse !== null) {
      this.fuse.add(author)
      return this.fuse
    }
    throw 'fuse not initialize yet'
  }

  /**
   * do initialize only if local fuse is null
   */
  async init(): Promise<Fuse<Author>> {
    if (this.authors === null || this.fuse === null) {
      const authors = await AuthorModel.getAll()
      const fuse = new Fuse(authors, {
        keys: ['name'],
      })
      this.authors = authors
      this.fuse = fuse
      return fuse
    }
    return this.fuse
  }

  async search(term: string): Promise<Author[]> {
    const fuse = this.fuse ?? (await this.init())
    return fuse.search(term).map(e => e.item)
  }
}

// export const SearchAuthorService = new SearchAuthorServiceClass() // export for global use
