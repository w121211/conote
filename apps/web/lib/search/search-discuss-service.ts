import Fuse from 'fuse.js'
import { Discuss } from '@prisma/client'
import { DiscussModel } from '../models/discuss-model'

class SearchDiscussServiceClass {
  private discusses: Discuss[] | null = null
  private fuse: Fuse<Discuss> | null = null

  async initFuse(): Promise<Fuse<Discuss>> {
    if (this.discusses === null || this.fuse === null) {
      const discusses = await DiscussModel.getAll()
      const fuse = new Fuse(discusses, {
        includeScore: true,
        keys: ['title'],
      })
      this.discusses = discusses
      this.fuse = fuse
      return fuse
    }
    return this.fuse
  }

  async search(term: string): Promise<string[]> {
    const fuse = this.fuse ?? (await this.initFuse())
    return fuse.search(term).map(e => e.item.title)
  }
}

export const SearchDiscussService = new SearchDiscussServiceClass() // export for global use
