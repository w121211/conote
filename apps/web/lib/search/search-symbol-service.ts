import Fuse from 'fuse.js'
import { Sym, SymType } from '@prisma/client'
import { SymModel } from '../models/sym-model'

// let symbols: string[] | null = null
// let symbolFuse: Fuse<string>
// let topicFuseUpdatedAt: Date | null = null
// let tickerTitles: string[] | null = null

// const prisma = new PA.PrismaClient({
//   errorFormat: 'pretty',
//   log: ['query', 'info', 'warn'],
// })

type FuseDict = {
  all: Fuse<string>
  ticker: Fuse<string>
  topic: Fuse<string>
}

class SearchSymbolServiceClass {
  // TODO: store in redis instead

  private syms: Sym[] | null = null
  private fuseDict: FuseDict | null = null

  async initFuse(): Promise<FuseDict> {
    if (this.syms === null || this.fuseDict === null) {
      const syms = await SymModel.getAll()

      // const fuse = new Fuse(syms, {
      //   includeScore: true,
      //   keys: ['symbol'],
      // })
      const fuseDict: FuseDict = {
        all: new Fuse(syms.filter(e => ['TICKER', 'TOPIC'].includes(e.type)).map(e => e.symbol)),
        ticker: new Fuse(syms.filter(e => e.type === 'TICKER').map(e => e.symbol)),
        topic: new Fuse(syms.filter(e => e.type === 'TOPIC').map(e => e.symbol)),
      }
      this.syms = syms
      this.fuseDict = fuseDict
      return fuseDict
    }
    return this.fuseDict
  }

  // async insertSymbol(sym: Sym) {
  // }

  async search(term: string, type?: SymType): Promise<string[]> {
    const fuseDict = this.fuseDict ?? (await this.initFuse())
    let fuse: Fuse<string>
    if (type === 'TICKER') {
      fuse = fuseDict['ticker']
    } else if (type === 'TOPIC') {
      fuse = fuseDict['topic']
    } else {
      fuse = fuseDict['all']
    }
    return fuse.search(term).map(e => e.item)
  }
}

// export for global use
export const SearchSymbolService = new SearchSymbolServiceClass()
