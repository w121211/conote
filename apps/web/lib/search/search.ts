import { Author, Discuss } from '@prisma/client'
import { AuthorModel } from '../models/author-model'
import { DiscussModel } from '../models/discuss-model'
import { SearchServiceClass } from './search-service'
import { SearchSymServiceClass } from './search-sym-service'

const searchAuthorServicePropertyName = `__prevent-name-collision__search-author-service`
const searchDiscussServicePropertyName = `__prevent-name-collision__search-discuss-service`
const searchSymServicePropertyName = `__prevent-name-collision__search-sym-service`

type GlobalThisWithSearchServices = typeof globalThis & {
  [searchAuthorServicePropertyName]: SearchServiceClass<Author>
  [searchDiscussServicePropertyName]: SearchServiceClass<Discuss>
  [searchSymServicePropertyName]: SearchSymServiceClass
}

const getSearchAuthorService = (): SearchServiceClass<Author> => {
  if (process.env.NODE_ENV === 'production') {
    return new SearchServiceClass<Author>({ keys: ['name'] }, AuthorModel.getAll)
  } else {
    const newGlobalThis = globalThis as GlobalThisWithSearchServices
    if (!newGlobalThis[searchAuthorServicePropertyName]) {
      console.log('Creating global search-author-service')
      newGlobalThis[searchAuthorServicePropertyName] = new SearchServiceClass<Author>(
        { keys: ['name'] },
        AuthorModel.getAll,
      )
    }
    return newGlobalThis[searchAuthorServicePropertyName]
  }
}

const getSearchDiscussService = (): SearchServiceClass<Discuss> => {
  if (process.env.NODE_ENV === 'production') {
    return new SearchServiceClass<Discuss>({ keys: ['title'] }, DiscussModel.getAll)
  } else {
    const newGlobalThis = globalThis as GlobalThisWithSearchServices
    if (!newGlobalThis[searchDiscussServicePropertyName]) {
      console.log('Creating global search-discuss-service')
      newGlobalThis[searchDiscussServicePropertyName] = new SearchServiceClass<Discuss>(
        { keys: ['title'] },
        DiscussModel.getAll,
      )
    }
    return newGlobalThis[searchDiscussServicePropertyName]
  }
}

const getSearchSymService = (): SearchSymServiceClass => {
  if (process.env.NODE_ENV === 'production') {
    return new SearchSymServiceClass()
  } else {
    const newGlobalThis = globalThis as GlobalThisWithSearchServices
    if (!newGlobalThis[searchSymServicePropertyName]) {
      console.log('Creating global search-sym-service')
      newGlobalThis[searchSymServicePropertyName] = new SearchSymServiceClass()
    }
    return newGlobalThis[searchSymServicePropertyName]
  }
}

export const SearchAuthorService = getSearchAuthorService()

export const SearchDiscussService = getSearchDiscussService()

export const SearchSymService = getSearchSymService()
