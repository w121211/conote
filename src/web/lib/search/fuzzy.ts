/**
 * Fuzzy search for page title, ticker, etc （用memory實現）
 */
// import _ from 'lodash'
// import dayjs from 'dayjs'
import { Card, CardType } from '@prisma/client'
import Fuse from 'fuse.js'
// import { Symbol as PrismaSymbol } from '@prisma/client'
import prisma from '../prisma'

// 將資料庫取出的資料存在記憶體中（定時更新），用於fuzzy search
// TODO: 改存在Redis
let symbols: string[] | null = null
let symbolFuse: Fuse<string>
// let topicFuseUpdatedAt: Date | null = null
// let tickerTitles: string[] | null = null

// const prisma = new PA.PrismaClient({
//   errorFormat: 'pretty',
//   log: ['query', 'info', 'warn'],
// })

// async function getAllTickers() {
//   const titles: string[] = [];
//   let cursor: number | null = null;
//   // eslint-disable-next-line no-constant-condition
//   while (true) {
//     let results;
//     if (cursor !== null) {
//       results = await prisma.page.findMany({
//         take: 10,
//         where: { template: { equals: 'TICKER' } },
//         orderBy: { id: 'asc' },
//         cursor: { id: cursor },
//       });
//     } else {
//       results = await prisma.page.findMany({
//         take: 10,
//         where: { template: { equals: 'TICKER' } },
//         orderBy: { id: 'asc' },
//       });
//     }
//     if (results.length === 0) break;
//     titles.concat(results.map(e => e.title));
//     cursor = results[-1].id;
//   }
//   return titles;
// }

export async function getAllSymbols(): Promise<string[]> {
  console.log('Loading symbols from database...')

  let symbols: string[] = []
  let cursor: number | undefined = undefined

  // eslint-disable-next-line no-constant-condition
  while (true) {
    // eslint-disable-next-line no-await-in-loop
    const res: Card[] = await prisma.card.findMany({
      // select: { symbol: true},
      where: {
        OR: [{ type: CardType.TICKER }, { type: CardType.TOPIC }],
      },
      take: 100,
      // skip cursor
      skip: cursor ? 1 : undefined,
      orderBy: { id: 'asc' },
      cursor: cursor ? { id: cursor } : undefined,
    })
    if (res.length === 0) {
      break
    }
    symbols = symbols.concat(res.map(e => e.symbol))
    cursor = res[res.length - 1].id
  }
  return symbols
}

export async function searchAllSymbols(term: string): Promise<string[]> {
  if (symbols === null) {
    symbols = await getAllSymbols()
    // console.log(symbols)
    symbolFuse = new Fuse(symbols, {
      includeScore: true,
      // keys: ['name'],
    })
  }
  // return symbolFuse.search(term).map(e => e.item.name)
  return symbolFuse.search(term).map(e => e.item)
}
