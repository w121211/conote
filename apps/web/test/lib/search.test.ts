/**
 * yarn run test-lib search.test
 */
/* eslint-disable no-await-in-loop */
import { SearchAuthorService, SearchSymService } from '../../lib/search/search'

it('SearchSymService.search() all', async () => {
  expect((await SearchSymService.search('AI')).map(e => e.symbol)).toMatchSnapshot()
  expect((await SearchSymService.search('半導體')).map(e => e.symbol)).toMatchSnapshot()
})

it('SearchSymService.search() ticker only', async () => {
  expect((await SearchSymService.search('ai', 'TICKER')).map(e => e.symbol)).toMatchSnapshot()
})

it('SearchSymService.search() topic only', async () => {
  expect((await SearchSymService.search('半導體', 'TOPIC')).map(e => e.symbol)).toMatchSnapshot()
})

it('SearchAuthorService.search()', async () => {
  expect((await SearchAuthorService.search('NaNa')).map(e => e.name)).toMatchSnapshot()
  expect((await SearchAuthorService.search('美股')).map(e => e.name)).toMatchSnapshot()
  // expect(await SearchSymbolService.search('半導體', 'TOPIC')).toMatchSnapshot()
})
