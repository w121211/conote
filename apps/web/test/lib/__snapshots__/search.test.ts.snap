// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchAuthorService.search() 1`] = `
Array [
  "NaNa说美股",
  "贝奇说股Beckie_Analysis",
  "Money_Manta的投资频道",
]
`;

exports[`SearchAuthorService.search() 2`] = `
Array [
  "老汤美股财经",
  "美投讲美股",
  "非專業投資美股",
  "ERay说美股",
  "NaNa说美股",
  "贝奇说股Beckie_Analysis",
]
`;

exports[`SearchSymService.search() all 1`] = `
Array [
  "$AI",
  "$ICE",
  "$IPOE",
  "$WISH",
  "$ABBV",
  "$ABNB",
  "$ADBE",
  "$AAPL",
  "$DIS",
  "$NIO",
  "$COIN",
  "[[PIPE]]",
  "[[aaa]]",
  "$NUAN",
  "$NDAQ",
  "$DMYI",
  "[[Unit]]",
  "[[Warrant]]",
  "[[pointbet]]",
  "$UFPI",
  "[[SPAC]]",
  "[[SoFi]]",
  "[[Bill Miller]]",
]
`;

exports[`SearchSymService.search() all 2`] = `
Array [
  "[[半導體]]",
]
`;

exports[`SearchSymService.search() ticker only 1`] = `
Array [
  "$AI",
  "$ICE",
  "$IPOE",
  "$WISH",
  "$ABBV",
  "$ABNB",
  "$ADBE",
  "$AAPL",
  "$DIS",
  "$NIO",
  "$COIN",
  "$NUAN",
  "$NDAQ",
  "$DMYI",
  "$UFPI",
]
`;

exports[`SearchSymService.search() topic only 1`] = `
Array [
  "[[半導體]]",
]
`;
